app/src/editor/shared/components/modules/Scheduler.jsx

```diff
@@ -0,0 +1,714 @@
+                )}
+                {rule.intervalType === 1 && (
+                    <DayControl onChange={this.onUpdateDates} {...rule} />
+                )}
+                {rule.intervalType === 2 && (
+                    <WeekControl onChange={this.onUpdateDates} {...rule} />
+                )}
+                {rule.intervalType === 3 && (
+                    <MonthControl onChange={this.onUpdateDates} {...rule} />
+                )}
+                {rule.intervalType === 4 && (
+                    <YearControl onChange={this.onUpdateDates} {...rule} />
+                )}
+            </div>
+        )
+    }
+}
+
+const Rule = withHover(RuleComponent)
+
+const defaultRule = {
+    intervalType: 0,
+    value: 0,
+    endDate: {
+        minute: 1,
+    },
+    startDate: {
+        minute: 0,
+    },
+}
+
+const addRule = (rules, id = uuid()) => ([
+    ...rules,
+    {
+        id,
+        ...defaultRule,
+    },
+])
+
+const removeRule = (rules, id) => rules.filter((rule) => rule.id !== id)
+
+const changeRule = (rules, id, properties) => rules.map((r) => {
+    if (r.id === id) {
+        return {
+            ...r,
+            ...properties,
+        }
+    }
+
+    return r
+})
+
+export default class SchedulerModule extends React.Component {
+    subscription = React.createRef()
+
+    state = {
+        defaultValue: undefined,
+        rules: undefined,
+    }
+
+    static getDerivedStateFromProps(props, state) {
+        const checksum = JSON.stringify(props.module.schedule.rules)
+        if (state.rules && state.checksum === checksum) {
+            return null
+        }
+
+        const { rules, defaultValue } = props.module.schedule || {}
+
+        // Add default row when module is added to canvas
+        const existingRules = rules || []
+        const defaultRules = (!existingRules.length && !state.rules) ? [{
+            ...defaultRule,
+        }] : existingRules
+
+        // Map existing rules with an id
+        return {
+            defaultValue: defaultValue || 0,
+            rules: defaultRules.map((rule) => ({
+                ...rule,
+                id: rule.id || uuid(),
+            })),
+            checksum,
+        }
+    }
+
+    onAddRule = () => {
+        this.setState((prevState) => ({
+            rules: addRule(prevState.rules),
+        }), this.updateStateToModule)
+    }
+
+    onRemoveRule = (id) => {
+        this.setState((prevState) => ({
+            rules: removeRule(prevState.rules, id),
+        }), this.updateStateToModule)
+    }
+
+    onChangeRule = (id, properties) => {
+        this.setState((prevState) => ({
+            rules: changeRule(prevState.rules, id, properties),
+        }), this.updateStateToModule)
+    }
+
+    onDefaultValueChange = (defaultValue) => {
+        this.setState({
+            defaultValue,
+        }, this.updateStateToModule)
+    }
+
+    updateStateToModule = () => {
```

Comment:
An alternative to putting this in every setState is to call it in componentDidUpdate. Slightly different but similar.

Reply:
Left as is for now since it works - I tried it but it didn't immediately work and so I decided not to investigate it deeper.